### PR TITLE
Use Heap To Maintain First Differences

### PR DESCRIPTION
**Pull request by jcastrence**
Signed-off-by: Julian Castrence <juliancastrence@ibm.com>

#### Type of change

- Improvement (improvement to code, performance, etc)

#### Description

- Changed ledgerutil compare naive implementation of maintaining list of first differences to use heap implementation for better performance

### CODE REVIEW

File: internal/ledgerutil/compare.go [hunk 1]
```diff
@@ -283,51 +282,63 @@ func findAndWriteDifferences(outputDirPath string, outputFilename string, snapsh
...
 
-func (s diffRecordSlice) Swap(i, j int) {
+func (s diffRecordHeap) Swap(i, j int) {
 	s[i], s[j] = s[j], s[i]
 }
 
-func (s diffRecordSlice) Less(i, j int) bool {
-	return (s[i]).lessThan(s[j])
+func (s diffRecordHeap) Less(i, j int) bool {
+	return !(s[i]).lessThan(s[j])
```
Comment (manish-sethi) [inline internal/ledgerutil/compare.go]: Just a suggestion to improve the readability here -- Reverse the semantics of `lessThan` and rename it to `committedAfter` and remove the negation in front.

File: internal/ledgerutil/compare.go [hunk 2]
```diff
@@ -283,51 +282,63 @@ func findAndWriteDifferences(outputDirPath string, outputFilename string, snapsh
...
-			s.highestRecord = r
-			s.highestIndex = i
-		}
+// Returns list of diffRecords currently in heap sorted from earliest to latest, will empty the heap
+func (s *firstRecords) getAllRecords() []*diffRecord {
+	n := len(*s.records)
+	var t []*diffRecord
+	for i := 0; i < n; i++ {
+		r := heap.Pop(s.records)
+		t = append([]*diffRecord{r.(*diffRecord)}, t...)
```
Comment (manish-sethi) [inline internal/ledgerutil/compare.go]: As you know the size in advance, you can avoid repetitive memory allocation. 
```suggestion
		t := make([]*diffRecord, n)
	for i := 0; i < n; i++ {
		r := heap.Pop(s.records)
		t[n-i-1] = r.(*diffRecord)
	}
```

**Review by manish-sethi — COMMENTED** [review]
LGTM, a couple minor comments

File: internal/ledgerutil/compare.go (hunk 2 above)
Comment (jcastrence) [reply internal/ledgerutil/compare.go]: Makes sense 👍🏼

File: internal/ledgerutil/compare.go (hunk 1 above)
Comment (jcastrence) [reply internal/ledgerutil/compare.go]: I changed `lessThan` to `earlierThan` which I feel makes a lot more sense and is more readable without requiring the semantics to be changed

**Review by manish-sethi — APPROVED** [review]

**Pull request merged**